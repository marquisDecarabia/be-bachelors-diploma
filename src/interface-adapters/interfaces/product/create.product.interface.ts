export interface CreateProduct {
  name: string;
  price: number;
  currency: string;
}
