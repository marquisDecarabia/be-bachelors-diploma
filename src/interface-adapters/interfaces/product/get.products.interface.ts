export interface GetProducts {
  limit: number;
  page: number;
}
