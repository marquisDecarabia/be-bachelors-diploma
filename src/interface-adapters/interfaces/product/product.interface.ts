import { ModelBase } from '@libs/ddd/interface-adapters/interfaces/model.base.interface';

export interface Product extends ModelBase {
  name: string;
  price: number;
  status: string;
  currency: string;
}
