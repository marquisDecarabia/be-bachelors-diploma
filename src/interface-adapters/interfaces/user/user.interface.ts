import { ModelBase } from '@libs/ddd/interface-adapters/interfaces/model.base.interface';

export interface User extends ModelBase {
  name: string;
}
