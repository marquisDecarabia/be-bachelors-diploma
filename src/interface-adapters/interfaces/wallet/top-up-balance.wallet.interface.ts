export interface TopUpBalanceWallet {
  walletId: string;
  amount: number;
}
