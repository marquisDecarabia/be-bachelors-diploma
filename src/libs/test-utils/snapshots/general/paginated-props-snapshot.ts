export const paginatedPropsSnapshot = {
  data: expect.any(Array),
  count: expect.any(Number),
  limit: expect.any(Number),
  page: expect.any(Number),
};
