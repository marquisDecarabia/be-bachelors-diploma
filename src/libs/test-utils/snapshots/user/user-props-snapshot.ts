export const userPropsSnapshot = {
  id: expect.any(String),
  createdAt: expect.any(String),
  updatedAt: expect.any(String),
  name: expect.any(String),
};
