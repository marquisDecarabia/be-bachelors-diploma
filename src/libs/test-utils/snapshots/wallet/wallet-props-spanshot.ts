export const walletPropsSnapshot = {
  id: expect.any(String),
  createdAt: expect.any(String),
  updatedAt: expect.any(String),
  userId: expect.any(String),
  balance: expect.any(Number),
  currency: expect.any(String),
};

export const walletPropsSnapshotWithParams = (
  userId: string,
  balanceNum: number,
  currency: string,
) => {
  return {
    ...walletPropsSnapshot,
    userId: userId,
    balance: balanceNum,
    currency: currency,
  };
};
