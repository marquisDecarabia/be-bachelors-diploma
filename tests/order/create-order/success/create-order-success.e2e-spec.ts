import { defineFeature, loadFeature } from 'jest-cucumber';
import * as request from 'supertest';
import { getTestServer, TestServer } from '../../../jestSetupAfterEnv';
import { Id } from '@libs/ddd/interface-adapters/interfaces/id.interface';
import { Product } from '@src/interface-adapters/interfaces/product/product.interface';
import { userPropsSnapshot } from '@libs/test-utils/snapshots/user/user-props-snapshot';
import { paginatedPropsSnapshot } from '@libs/test-utils/snapshots/general/paginated-props-snapshot';
import {
  productPropsSnapshot,
  productPropsSnapshotWithParams,
} from '@libs/test-utils/snapshots/product/product-props-snapshot';
import { walletPropsSnapshotWithParams } from '@libs/test-utils/snapshots/wallet/wallet-props-spanshot';

const feature = loadFeature(
  'tests/order/create-order/success/create-order-success.feature',
);

defineFeature(feature, (test) => {
  let testServer: TestServer;
  let httpServer: request.SuperTest<request.Test>;

  beforeAll(() => {
    testServer = getTestServer();
    httpServer = request(testServer.serverApplication.getHttpServer());
  });

  afterAll(() => {
    // TODO: clean db after tests are finished OR reinitialize db w/ seed
  });

  test('Creating an order (success)', ({ given, when, then, and }) => {
    let userId: Id;
    let products: Product[];
    let product: Product;
    let orderId: Id;

    given(/^that my user ID is "(.*)"$/, async (id: string) => {
      const result = await httpServer.get(`/v1/users/${id}`).send().expect(200);
      expect(result.body).toMatchSnapshot(userPropsSnapshot);
      userId = { id };
    });

    and(
      /^my wallet balance is "(.*)" "(.*)"$/,
      async (balance: string, currency: string) => {
        const numBalance = Number(balance);
        const res = await httpServer
          .get(`/v1/wallets/user/${userId.id}`)
          .send()
          .expect(200);
        expect(res.body).toMatchSnapshot(
          walletPropsSnapshotWithParams(userId.id, numBalance, currency),
        );
      },
    );

    when('i request products list', async () => {
      const res = await httpServer
        .get('/v1/products?page=1&limit=20')
        .send()
        .expect(200);
      expect(res.body).toMatchSnapshot(paginatedPropsSnapshot);
      res.body.data.forEach((product) => {
        expect(product).toMatchSnapshot(productPropsSnapshot);
      });
      products = res.body.data;
    });

    then(
      /^I see product with name is "(.*)" and price is (\d+) "(.*)" and status is "(.*)"$/,
      (productName, price, currency, status) => {
        const numPrice = Number(price);
        product = products.find((item) => {
          return (
            item.name === productName &&
            item.price === numPrice &&
            item.currency === currency &&
            item.status === status
          );
        });
        expect(product).toMatchSnapshot(productPropsSnapshot);
      },
    );

    when('I send a request to create an order with this product', async () => {
      const res = await httpServer
        .post('/v1/orders')
        .send({
          userId: userId.id,
          productId: product.id,
        })
        .expect(201);
      orderId = res.body;
    });

    then('I receive order ID', () => {
      expect(orderId).toMatchSnapshot({
        id: expect.any(String),
      });
    });

    // and('I can see order details by this ID', () => {
    //   // const res = await httpServer.get('v1/orders')
    // });

    and(/^product status is "(.*)"$/, async (status) => {
      const res = await httpServer
        .get('/v1/products?page=1&limit=20')
        .send()
        .expect(200);
      const updatedProduct = res.body.data.find(
        (item) => item.id === product.id,
      );

      expect(updatedProduct).toMatchSnapshot(
        productPropsSnapshotWithParams(
          product.name,
          product.price,
          product.currency,
          status,
        ),
      );
    });

    and(/^my wallet balance is "(.*)" "(.*)"$/, async (balance, currency) => {
      const numBalance = Number(balance);
      const res = await httpServer
        .get(`/v1/wallets/user/${userId.id}`)
        .send()
        .expect(200);
      expect(res.body).toMatchSnapshot(
        walletPropsSnapshotWithParams(userId.id, numBalance, currency),
      );
    });
  });
});
