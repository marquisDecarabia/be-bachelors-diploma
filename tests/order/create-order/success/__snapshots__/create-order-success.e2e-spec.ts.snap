// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create a order Creating an order (success) 1`] = `
Object {
  "createdAt": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create a order Creating an order (success) 2`] = `
Object {
  "balance": 217.35,
  "createdAt": Any<String>,
  "currency": "USD",
  "id": Any<String>,
  "updatedAt": Any<String>,
  "userId": "7e0154e2-aca2-4af6-92b4-bd5c517f5b36",
}
`;

exports[`Create a order Creating an order (success) 3`] = `
Object {
  "count": Any<Number>,
  "data": Any<Array>,
  "limit": Any<Number>,
  "page": Any<Number>,
}
`;

exports[`Create a order Creating an order (success) 4`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create a order Creating an order (success) 5`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create a order Creating an order (success) 6`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create a order Creating an order (success) 7`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create a order Creating an order (success) 8`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create a order Creating an order (success) 9`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create a order Creating an order (success) 10`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create a order Creating an order (success) 11`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create a order Creating an order (success) 12`] = `
Object {
  "createdAt": Any<String>,
  "currency": Any<String>,
  "id": Any<String>,
  "name": Any<String>,
  "price": Any<Number>,
  "status": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`Create a order Creating an order (success) 13`] = `
Object {
  "id": Any<String>,
}
`;

exports[`Create a order Creating an order (success) 14`] = `
Object {
  "createdAt": Any<String>,
  "currency": "USD",
  "id": Any<String>,
  "name": "Lava Lake",
  "price": 150,
  "status": "sold",
  "updatedAt": Any<String>,
}
`;

exports[`Create a order Creating an order (success) 15`] = `
Object {
  "balance": 67.35,
  "createdAt": Any<String>,
  "currency": "USD",
  "id": Any<String>,
  "updatedAt": Any<String>,
  "userId": "7e0154e2-aca2-4af6-92b4-bd5c517f5b36",
}
`;
